File: mqtt-event-logger/package/mqtt-char-driver/src/mqtt_char_driver.rs
// SPDX-License-Identifier: GPL-2.0

#![no_std]

use kernel::prelude::*;

module! {
    type: MQTTCharDriver,
    name: "mqtt_char_driver",
    author: "Cornelia Schulz",
    description: "Event logging kernel module",
    license: "GPL",
}

struct MQTTCharDriver;

impl kernel::Module for MQTTCharDriver {
    fn init(_module: &'static ThisModule) -> Result<Self> {
        pr_info!("mqtt_char_driver: module loaded\n");

        Ok(MQTTCharDriver)
    }
}

impl Drop for MQTTCharDriver {
    fn drop(&mut self) {
        pr_info!("mqtt_char_driver: module unloaded\n");
    }
}

File: mqtt-event-logger/package/mqtt-char-driver/src/mqtt_char_driver.c
// SPDX-License-Identifier: GPL-2.0

#include <linux/init.h>
#include <linux/kernel.h>
#include <linux/module.h>

MODULE_DESCRIPTION("MQTT Character Driver");
MODULE_AUTHOR("Cornelia Schulz");
MODULE_VERSION("1.0.0");
MODULE_LICENSE("GPL");

static int __init mqtt_char_driver_init(void)
{
    pr_info("mqtt_char_driver: module loaded\n");
    return 0;
}

static void __exit mqtt_char_driver_exit(void)
{
    pr_info("mqtt_char_driver: module unloaded\n");
}

module_init(mqtt_char_driver_init);
module_exit(mqtt_char_driver_exit);